feat: add cross region support for buckets

jor2 opened the pull request:
### Description

CRA scan is failing with Goal ID 3000116 : Check whether Cloud Object Storage bucket resiliency is set to cross region

For this to pass, I think we need to add cross region support to https://github.com/terraform-ibm-modules/terraform-ibm-cos

The ibm_cos_bucket terraform resource supports:

cross_region_location - (Optional, String) Specify the cross-regional bucket location. Supported values are us, eu, and ap. If you use this parameter, do not set single_site_location or region_location at the same time.

so we just need to expose this in the module.

**Check the relevant boxes:**
- [ ] Bug fix (nonbreaking change that fixes an issue)
- [x] New feature (nonbreaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI related update (pipeline, etc.)

### Checklist

- [ ] If relevant, a test for the change has been added or updated as part of this PR.
- [ ] If relevant, documentation for the change has been added or updated as part of this PR.

### Merge

- Merge using "Squash and merge".
- Make sure to use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents. The commit message determines whether a new version of the modules needs to be released, and if so, which semver number to use).

ocofaigh reviewed (changes requested):
`cross_region_location` need to be exposed as an input variable so consumers can select the value they want. According to the [doc](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/cos_bucket), Supported values are `us`, `eu`, and `ap` so lets add validation for that.
It also says if you use this parameter, do not set `single_site_location` or `region_location` at the same time, so we need to handle that too because the code in this PR is setting values for both right now

ocofaigh commented:
I think we should also create a new example that will show how to use cross region as apposed to a single region, and run it as part of pr_test.go
We should then be able to update the CRA scan to scan that example, and remove the `3000116` goal from the CRA ignore list json file

shemau commented:
I think passing a property is insufficient. If you are lucky you might be able to force region_location to null (when setting cross_region_location). Also force cross_region_location to null when setting region_location.
Otherwise, like with key_protect; it may be necessary to create two more bucket definitions. Similar model, instead of 1 of 2 being used, it becomes 1 of 4 and the right id/name/public/private have to be set in the locals (L188) so that outputs work as expected.

ocofaigh commented:
I think we can pass null value for one or the other, but we will either need a new variable to say whether to do single region or cross region, or else some validation which fails if a value is passed for both

jor2 commented:
> I think we can pass null value for one or the other, but we will either need a new variable to say whether to do single region or cross region, or else some validation which fails if a value is passed for both

I've added some validation to limit the number of values you can pass to `cross_region_location`. Also added some validation in the locals to verify both values `region` and `cross_region_location` cannot be set at the same time. I'm still quite new to terraform so maybe this isn't sufficient enough.

ocofaigh commented on variables.tf:
```diff
@@ -67,6 +67,17 @@ variable "create_cos_bucket" {
 }
 
+variable "cross_region_location" {
+  description = "Specify the cross-regional bucket location. Supported values are 'us', 'eu', and 'ap'."
```
Specify the cross-regional bucket location. Supported values are 'us', 'eu', and 'ap'. If you pass a value for this, do not pass one for var.region

ocofaigh reviewed (changes requested):
@jor2 Can we update the description of the `region` variable (its actually currently incorrect in main branch):
```
The region to provision the bucket. If you pass a value for this, do not pass one for var.cross_region_location
```

Instead of actually creating a new example, we could update the example in `examples/complete` so that `cos_bucket2` actually provisions a cross region bucket instead. That way the test will test both. Ensure to update the readme of the example then too

ocofaigh commented on examples/complete/variables.tf:
```diff
@@ -22,6 +22,17 @@ variable "region" {
 }
 
+variable "cross_region_location" {
+  description = "Specify the cross-regional bucket location. Supported values are 'us', 'eu', and 'ap'. If you pass a value for this, do not pass one for var.region"
```
"If you pass a value for this, do not pass one for var.region"

^^ Remove this from the example variable description. In our example, we will need both variables to be passed, as region is used to provision other things such as COS instance, key protect and observability instances

ocofaigh commented on variables.tf:
```diff
@@ -77,6 +77,23 @@ variable "create_cos_bucket" {
+}
+
+variable "cross_region_location" {
+  description = "Specify the cross-regional bucket location. Supported values are 'us', 'eu', and 'ap'. If you pass a value for this, do not pass one for var.region."
```
Actually, might be clearer if said "If you pass a value for this, ensure to set the value of var.region to null"

ocofaigh commented on variables.tf:
```diff
@@ -77,6 +77,23 @@ variable "create_cos_bucket" {
   default     = true
 }
 
+variable "cross_region_location_enabled" {
```
Do we really need this extra variable? I think we can remove it and just have validation logic that would fail if values for var.region and var.cross_region_location were both passed

ocofaigh commented on main.tf:
```diff
@@ -25,6 +25,8 @@ locals {
   # tflint-ignore: terraform_unused_declarations
   validate_kp_guid_input = var.encryption_enabled && var.create_cos_instance && var.existing_key_protect_instance_guid == null ? tobool("A value must be passed for var.existing_key_protect_instance_guid when var.create_cos_instance and var.encryption_enabled is true.") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_cross_region_location_inputs = var.cross_region_location_enabled && var.cross_region_location == null ? tobool("If var.cross_region_location_enabled is true, then provide value for var.cross_region_location") : true
```
I think we should update this to:

`validate_cross_region_location_inputs = var.create_cos_bucket && ((var.cross_region_location == null && var.region == null) || (var.cross_region_location != null && var.region != null)) ? tobool("If var.create_cos_bucket is true, then value needs to be provided for var.cross_region_location or var.region, but not both") : true`

ocofaigh reviewed (changes requested):
left some comments

ocofaigh commented:
@jor2 don't commit updates to the `common-dev-assets` git submodule, or you will end up with conflicts very quickly (like you have now) due to the renovate automation we have running several times a day that bumps in in main branch. Suggest removing it from this PR

jor2 commented:
> @jor2 don't commit updates to the `common-dev-assets` git submodule, or you will end up with conflicts very quickly (like you have now) due to the renovate automation we have running several times a day that bumps in in main branch. Suggest removing it from this PR

makes sense, those conflicts are very annoying 😠.

ocofaigh commented on tests/pr_test.go:
```diff
@@ -32,10 +32,10 @@ func setupOptions(t *testing.T, prefix string, dir string) *testhelper.TestOptio
 		TerraformDir:                  dir,
 		Prefix:                        prefix,
 		ResourceGroup:                 resourceGroup,
+		Region:                        "us-south",
```
add a comment here:
// lock into us-south as is supports cross region support when using your own KMS key

ocofaigh replied:
@jor2 comment still missing here

ocofaigh commented:
Hmm, the test failed with:
`Error: failed to update the lifecyle rule on COS bucket cos-complete-2nsjgr-bucket-2, InvalidRequest: The operation requested is not supported`

I dont know if this is another feature that is unsupported with cross region? Were you able to successfully run this locally?

ocofaigh commented on main.tf:
```diff
@@ -25,6 +25,10 @@ locals {
+  # tflint-ignore: terraform_unused_declarations
+  validate_cross_region_location_inputs = var.create_cos_bucket && ((var.cross_region_location == null && var.region == null) || (var.cross_region_location != null && var.region != null)) ? tobool("If var.create_cos_bucket is true, then value needs to be provided for var.cross_region_location or var.region, but not both") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_cross_region_location_archive_disabled_inputs = var.cross_region_location != null && var.archive_days != null ? tobool("If var.cross_region_location is set, then var.expire_days cannot be set.") : true
```
this only applies if `var.create_cos_bucket` is true, so update the logic to reflect this (same way we have in `validate_cross_region_location_inputs`)

ocofaigh commented on variables.tf:
```diff
@@ -150,7 +161,7 @@ variable "object_versioning_enabled" {
 
 variable "archive_days" {
-  description = "Specifies the number of days when the archive rule action takes effect. Only used if 'create_cos_bucket' is true."
+  description = "Specifies the number of days when the archive rule action takes effect. Only used if 'create_cos_bucket' is true. Cannot be set when using var.cross_region_location as archive data is not supported with this feature."
```
must be set to `null` when when using var.cross_region_location as archive data is not supported with this feature.

ocofaigh reviewed (changes requested):
see latest comments

ocofaigh commented on examples/complete/main.tf:
```diff
@@ -13,21 +13,19 @@ module "resource_group" {
+  enable_platform_logs       = false
+  enable_platform_metrics    = false
+  logdna_provision           = false
+  activity_tracker_provision = false
```
the example still needs to support create a new activity tracker instance. So you need logic to mark this as true if no value is passed for `activity_tracker_crn`

ocofaigh commented on variables.tf:
```diff
@@ -8,9 +8,9 @@ variable "resource_group_id" {
+  description = "The region to provision the bucket. If you pass a value for this, do not pass one for var.cross_region_location."
   type        = string
-  default     = "us-south"
+  default     = null
```
I think we should still default this to a region so the default behaviour will be to deploy a single region bucket )otherwise its a breaking change for current consumers)

ocofaigh commented on variables.tf:
```diff
@@ -174,7 +185,7 @@ variable "expire_days" {
   type        = string
   description = "Activity tracker crn for COS bucket (Optional)"
-  default     = null
+  default     = "crn:v1:bluemix:public:logdnaat:eu-de:a/abac0df06b644a9cabc6e44f55b3880e:b1ef3365-dfbf-4d8f-8ac8-75f4f84d6f4a::"
```
default this to null. The CRN will be passed in pr_test.go

ocofaigh commented on variables.tf:
```diff
@@ -95,6 +95,17 @@ variable "create_cos_bucket" {
+variable "cross_region_location" {
+  description = "Specify the cross-regional bucket location. Supported values are 'us', 'eu', and 'ap'. If you pass a value for this, ensure to set the value of var.region to null."
+  type        = string
+  default     = "us"
```
default this to null, to allow regional bucket as the default setting

ocofaigh commented on tests/pr_test.go:
```diff
@@ -32,10 +34,10 @@ func setupOptions(t *testing.T, prefix string, dir string) *testhelper.TestOptio
 		CloudInfoService:              sharedInfoSvc,
 		ExcludeActivityTrackerRegions: true,
 	})
-
```
ExcludeActivityTrackerRegions can now be removed, we wont be deploying activity tracker anymore. It also means the `TestMain` function can be removed too

ocofaigh commented on tests/pr_test.go:
```diff
@@ -32,10 +34,10 @@ func setupOptions(t *testing.T, prefix string, dir string) *testhelper.TestOptio
 		TerraformDir:                  dir,
 		Prefix:                        prefix,
 		ResourceGroup:                 resourceGroup,
+		Region:                        region, // Not all regions provide cross region support so value must be hardcoded https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-service-availability.
```
you will ned to pass the CRN as a terraform var here:

```
TerraformVars: map[string]interface{}{
			"activity_tracker_crn":   crn.......,
		}
```

ocofaigh commented on examples/complete/variables.tf:
```diff
@@ -16,14 +16,32 @@ variable "resource_tags" {
+variable "activity_tracker_crn" {
+  type        = string
+  description = "Activity tracker crn for COS bucket. Using permanent instance as region is hardcoded in tests/pr_test.go."
+  default     = "crn:v1:bluemix:public:logdnaat:eu-de:a/abac0df06b644a9cabc6e44f55b3880e:b1ef3365-dfbf-4d8f-8ac8-75f4f84d6f4a::"
```
default this null (meaning a new AT instance will be created by the example). In the test will will pass a value for this

ocofaigh commented on examples/complete/main.tf:
```diff
@@ -61,28 +59,32 @@ module "cos_bucket1" {
   key_protect_key_crn                = module.key_protect_all_inclusive.keys["${local.key_ring_name}.${local.key_name}"].crn
   sysdig_crn                         = module.observability_instances.sysdig_crn
-  activity_tracker_crn               = module.observability_instances.activity_tracker_crn
+  activity_tracker_crn               = var.activity_tracker_crn
```
you will need a local variable now with ternary syntax to either use `var.activity_tracker_crn` value if passed in, or otherwise use the CRN of the AT instance that will be created by the example

ocofaigh reviewed (changes requested):
see comments

daniel-butler-irl commented:
@jor2 please try to avoid force push, should be used as a last resort. It can be unsafe because it rewrites the commit history. For more details see https://www.git-tower.com/blog/force-push-in-git/

ocofaigh commented on tests/pr_test.go:
```diff
@@ -16,26 +16,35 @@ const completeExistingTerraformDir = "examples/existing-resources"
 
-	os.Exit(m.Run())
-}
+var sharedInfoSvc *cloudinfo.CloudInfoService
```
This variable is not needed anymore - remove

ocofaigh commented on tests/pr_test.go:
```diff
@@ -16,26 +16,35 @@ const completeExistingTerraformDir = "examples/existing-resources"
+		Prefix:           prefix,
+		ResourceGroup:    resourceGroup,
+		Region:           region, // Not all regions provide cross region support so value must be hardcoded https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-service-availability.
+		CloudInfoService: sharedInfoSvc,
```
no longer needed - remove

daniel-butler-irl commented on .github/workflows/ci.yml:
```diff
@@ -18,3 +18,4 @@ jobs:
     with:
       craTarget: "examples/complete"
       craGoalIgnoreFile: "cra-tf-validate-ignore-goals.json"
+      craEnvironmentVariables: "TF_VAR_existing_at_instance_crn=crn:v1:bluemix:public:logdnaat:eu-de:a/abac0df06b644a9cabc6e44f55b3880e:b1ef3365-dfbf-4d8f-8ac8-75f4f84d6f4a::"
```
@ocofaigh do we want the at instance added as an org level variable so other repos can use it?

daniel-butler-irl replied:
I can put it in `PERMANENT_AT_INSTANCE_CRN_GE_DEV` 
Usage would be:
```
craEnvironmentVariables: "TF_VAR_existing_at_instance_crn={{ vars.PERMANENT_AT_INSTANCE_CRN_GE_DEV }}"
```

ocofaigh replied:
hmm, I already added it to https://github.com/terraform-ibm-modules/common-dev-assets/blob/main/common-go-assets/common-permanent-resources.yaml in which the tests will be updated to pull from (Jordan will add in subsequent PR). Is there any way to pull from there and set as org level variable so we only have to main in once place?

daniel-butler-irl replied:
Yes but would require dev work. Flow read file and update org variables with REST API on a cron or part of the merge pipeline for the source repo https://docs.github.com/en/rest/actions/variables?apiVersion=2022-11-28#create-an-organization-variable

ocofaigh replied:
Can't we just have the common pipeline execute some logic to pull from the yaml at runtime, since it comes from common-dev-assets which the consuming repo will have? (lets continue this conversation on slack as I am about the merge this PR)

ocofaigh commented on examples/complete/variables.tf:
```diff
@@ -16,14 +16,32 @@ variable "resource_tags" {
+
+variable "existing_at_instance_crn" {
+  type        = string
+  description = "Activity tracker crn for COS bucket. Using permanent instance as region is hardcoded in tests/pr_test.go."
```
update this to:
Optionally pass an existing activity tracker instance CRN to use in the example. If not passed, a new instance will be provisioned

ocofaigh commented on examples/complete/variables.tf:
```diff
@@ -16,14 +16,32 @@ variable "resource_tags" {
+
+  validation {
+    condition     = var.cross_region_location == null || can(regex("us|eu|ap", var.cross_region_location))
+    error_message = "Variable 'cross_region_location' must be 'us' or 'eu', 'ap', or 'null'."
```
for the sake of this example, null is not a valid option since we are deploying a cross region bucket

ocofaigh reviewed (changes requested):
see final minor comments

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 5.3.0 :tada:

The release is available on [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-cos/releases/tag/v5.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
